File: JavaScript Interview Concepts/Educative/Two Pointer/ReverseWordsInString.js
function reverseWords(sentence) {

    // Replace this placeholder return statement with your code
    let left=0,right=0,end=sentence.length;
    let newSentence="";
    // sentence.
    console.log(sentence.split(" ").filter((word)=>word!=="").map((word)=>word.split("").reverse().join(""))
                  .reverse().map((word)=> word.split("").reverse().join("")).join(" "))
    while(left<end){
        while(sentence.charAt(right) !== ' ' && right<end)
            right++;

        newSentence+=sentence.substring(left,right).split("").reverse().join("") + " "
        left = ++right;
        // console.log(newSentence);

    }
    return "";
}

reverseWords("We    Love Javascript      ")

File: Educative/Grooking the Coding Interview/Two Pointer/ReverseWordsInString.py
def reverse_words(sentence: str) -> str:
    # Replace this placeholder return statement with your code
    if len(sentence) == 0:
        return sentence

    slow = 0
    sentence = sentence[::-1]
    sentence = " ".join(sentence.split())
    i = 0
    while i < len(sentence):
        while i < len(sentence) and sentence[i] != " ":
            i += 1
        sentence = sentence[:slow] + sentence[slow:i][::-1] + sentence[i:]
        i += 1
        slow = i


    return sentence


# Test cases
if __name__ == "__main__":
    # Test case 1
    sentence1 = "hello world"
    print("Test case 1:", reverse_words(sentence1))  # Expected output: "world hello"

    # Test case 2
    sentence2 = "  this is   a test  "
    print("Test case 2:", reverse_words(sentence2))  # Expected output: "test a is this"

    # Test case 3
    sentence3 = "singleword"
    print("Test case 3:", reverse_words(sentence3))  # Expected output: "singleword"

    # Test case 4
    sentence4 = ""
    print("Test case 4:", reverse_words(sentence4))  # Expected output: ""
